storage/tests/python/src/norcow.py

```diff
@@ -103,64 +174,162 @@ def _append(self, key: int, value: bytes):
-        if key == consts.NORCOW_KEY_FREE:
+    def _read_item(self, offset: int, padding: bool = False) -> (int, int, bytes):
+        if offset >= consts.NORCOW_SECTOR_SIZE:
             raise ValueError("Norcow: no data on this offset")
-        length = self.sectors[self.active_sector][offset + 2 : offset + 4]
-        length = int.from_bytes(length, sys.byteorder)
-        value = self.sectors[self.active_sector][offset + 4 : offset + 4 + length]
-        return key, value
+
+        if self.flash_byte_access:
+            key = self.sectors[self.active_sector][offset : offset + 2]
+            key = int.from_bytes(key, sys.byteorder)
+            if key == consts.NORCOW_KEY_FREE:
+                raise ValueError("Norcow: no data on this offset")
+            length = self.sectors[self.active_sector][offset + 2 : offset + 4]
+            length = int.from_bytes(length, sys.byteorder)
+            value = self.sectors[self.active_sector][offset + 4 : offset + 4 + length]
+        else:
+
+            length = self.sectors[self.active_sector][offset : offset + 2]
+            length = int.from_bytes(length, sys.byteorder)
+
+            if length <= 12:
+                key = self.sectors[self.active_sector][offset + 2 : offset + 4]
+                key = int.from_bytes(key, sys.byteorder)
+                if key == consts.NORCOW_KEY_FREE:
+                    raise ValueError("Norcow: no data on this offset")
+                if padding:
+                    value = self.sectors[self.active_sector][
+                        offset : offset + align_int_add(4 + length, self.word_size)
+                    ]
+                else:
+                    value = self.sectors[self.active_sector][
+                        offset + 4 : offset + 4 + length
+                    ]
+            else:
+                key = self.sectors[self.active_sector][offset + 2 : offset + 4]
+                key = int.from_bytes(key, sys.byteorder)
+                if key == consts.NORCOW_KEY_FREE:
+                    raise ValueError("Norcow: no data on this offset")
+                deleted = self.sectors[self.active_sector][offset + self.word_size]
+                if padding:
+                    value = self.sectors[self.active_sector][
+                        offset : offset
+                        + align_int_add(self.word_size + 1 + length, self.word_size)
+                    ]
+                else:
+                    value = self.sectors[self.active_sector][
+                        offset
+                        + self.word_size
+                        + 1 : offset
+                        + self.word_size
+                        + 1
+                        + length
+                    ]
+                if deleted == 0:
+                    key = 0
+                else:
+                    if key == consts.NORCOW_KEY_FREE:
+                        raise ValueError("Norcow: no data on this offset")
+
+        return key, length, value
```

Comment:
What's the difference between `length` and `len(value)`?

Reply:
In `is_updatable` in `blockwise` implementation i use the `padding` param to specify that i want to read the whole item including padding, as i need to know the whole thing to be able to determine updatability. Then l`en(value)` is the lenght of the whole item and `length` is the length of data only

Reply:
In that case call it `padded_value`, because this is not at all obvious. However, as discussed elsewhere, I'd rather not deal with updatability in the blockwise implementation, because it's not worth the trouble.

Reply:
removed with the updatability simplification https://github.com/trezor/trezor-firmware/commit/f625f87fb56badc70fb6a47ff6ab4574aad0aed6